optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,480 @@
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.core.api.score.stream.Joiners;
+import org.optaplanner.core.api.score.stream.bi.BiConstraintStream;
+import org.optaplanner.core.api.score.stream.uni.UniConstraintStream;
+import org.optaplanner.core.impl.score.stream.drools.common.BiTuple;
+import org.optaplanner.core.impl.score.stream.drools.common.QuadTuple;
+import org.optaplanner.core.impl.util.Break;
+import org.optaplanner.core.impl.util.ConsecutiveData;
+import org.optaplanner.core.impl.util.Sequence;
+import org.optaplanner.examples.common.ExperimentalConstraintCollectors;
+import org.optaplanner.examples.nurserostering.domain.Employee;
+import org.optaplanner.examples.nurserostering.domain.NurseRosterParametrization;
+import org.optaplanner.examples.nurserostering.domain.ShiftAssignment;
+import org.optaplanner.examples.nurserostering.domain.ShiftDate;
+import org.optaplanner.examples.nurserostering.domain.ShiftType;
+import org.optaplanner.examples.nurserostering.domain.ShiftTypeSkillRequirement;
+import org.optaplanner.examples.nurserostering.domain.SkillProficiency;
+import org.optaplanner.examples.nurserostering.domain.contract.BooleanContractLine;
+import org.optaplanner.examples.nurserostering.domain.contract.ContractLine;
+import org.optaplanner.examples.nurserostering.domain.contract.ContractLineType;
+import org.optaplanner.examples.nurserostering.domain.contract.MinMaxContractLine;
+import org.optaplanner.examples.nurserostering.domain.contract.PatternContractLine;
+import org.optaplanner.examples.nurserostering.domain.pattern.FreeBefore2DaysWithAWorkDayPattern;
+import org.optaplanner.examples.nurserostering.domain.pattern.ShiftType2DaysPattern;
+import org.optaplanner.examples.nurserostering.domain.pattern.ShiftType3DaysPattern;
+import org.optaplanner.examples.nurserostering.domain.request.DayOffRequest;
+import org.optaplanner.examples.nurserostering.domain.request.DayOnRequest;
+import org.optaplanner.examples.nurserostering.domain.request.ShiftOffRequest;
+import org.optaplanner.examples.nurserostering.domain.request.ShiftOnRequest;
+import org.optaplanner.examples.nurserostering.score.drools.EmployeeConsecutiveAssignmentEnd;
+import org.optaplanner.examples.nurserostering.score.drools.EmployeeConsecutiveAssignmentStart;
+
+public class NurseRosteringConstraintProvider implements ConstraintProvider {
+
+    public <C extends ContractLine> BiConstraintStream<ImmutablePair<Employee, C>, Sequence<ShiftDate>>
+            getConsecutiveShifts(UniConstraintStream<C> constraintStream) {
+        return constraintStream.join(Employee.class, Joiners.equal(ContractLine::getContract, Employee::getContract))
+                .join(ShiftAssignment.class, Joiners.equal((contract, employee) -> employee, ShiftAssignment::getEmployee))
+                .groupBy((contract, employee, shift) -> ImmutablePair.of(employee, contract),
```

Comment:
I'd rather we not use `ImmutablePair` here if a suitable `groupBy()` overload is available.
It makes for a code that reads better, and does not introduce users to new types which are not strictly necessary.
It may even perform a bit better, as we do not create needless tuple objects all the time.

Reply:
It was ImmutablePair, but unfortunately ImmutablePair implements Comparable, meaning that when a score corruption occurs (eariler when I was debugging), it tries to access it compareTo method, which fails since MinMaxContractLine does not implement Comparable.